Remotion/Web/UnitTests/Core/Resources/ResourcePathBuilderBasedStaticResourceCacheKeyProviderTest.cs

```diff
@@ -0,0 +1,310 @@
+          resourcesPhysicalPathLocatorMock.Object,
+          new ResourcePathBuilderBasedStaticResourceCacheKeyProvider.ResourceFileDetailsAppender(),
+          resourceTypes);
+
+      Assert.That(testStaticResourceCacheKeyProvider.GetStaticResourceCacheKey(), Is.EqualTo(CreateExpectedCacheKeyForEntries()));
+
+      // Changes in the root folder should not be considered as long as they aren't deletions or renames
+      folder.CreateDirectory("Remotion.Web");
+      folder.CreateFile("abc.txt", "testContent", DateTime.Today, DateTime.Today);
+
+      // Changes to the assembly folders should not be considered as long as they aren't deletions or renames
+      folder.CreateDirectory(@"Remotion.Web\Html");
+      folder.CreateDirectory(@"Remotion.Web\UI");
+      folder.CreateFile(@"Remotion.Web\abc.txt", "testContent", DateTime.Today, DateTime.Today);
+
+      // Changes to relevant folders only consider files not directories
+      folder.CreateDirectory(@"Remotion.Web\Html\subfolder", DateTime.Today, DateTime.Today);
+
+      // Changes to irrelevant folders should be ignored completely
+      folder.CreateDirectory(@"Remotion.Web\UI\subfolder");
+      folder.CreateFile(@"Remotion.Web\UI\subfolder\abc.txt", "testContent", DateTime.Today, DateTime.Today);
+
+      // No better way than to sleep as we expect no notifications to come in and
+      // they can have an arbitrary delay. 100ms should be good enough as tests
+      // with 50ms also worked well.
+      Thread.Sleep(100);
+
+      Assert.That(testStaticResourceCacheKeyProvider.HasCacheKey, Is.True);
+    }
+
+    private void WaitForCacheKeyChange (TestStaticResourceCacheKeyProvider staticResourceCacheKeyProvider)
+    {
+      for (var i = 0; i < 10; i++)
+      {
+        if (!staticResourceCacheKeyProvider.HasCacheKey)
+          return;
+
+        Thread.Sleep(20);
+      }
+
+      throw new InvalidOperationException($"Expected cache key to change but it did not change within the timeout.");
+    }
+
+    private static string CreateExpectedCacheKeyForEntries (params (string Name, long Length, long CreatedTime, long LastWriteTime)[] entries)
+    {
+      var memoryStream = new MemoryStream();
+      using (var binaryWriter = new BinaryWriter(memoryStream, Encoding.UTF8, true))
+      {
+        foreach (var (name, length, creationTime, lastWriteTime) in entries)
+        {
+#if NETFRAMEWORK
+          binaryWriter.Write(name);
+#else
+          binaryWriter.Write(name.AsSpan());
+#endif
+          binaryWriter.Write(length);
+          binaryWriter.Write(creationTime);
+          binaryWriter.Write(lastWriteTime);
+        }
+      }
+
+      memoryStream.Position = 0;
+
+      using var sha256 = SHA256.Create();
+      var data = sha256.ComputeHash(memoryStream);
+
+#if NETFRAMEWORK
+      return string.Concat(data.Select(e => e.ToString("X2")));
+#else
+      return Convert.ToHexString(data);
+#endif
+    }
```

Comment:
I'm not sure this is such a great idea / helpful. We just copied the algorithm so any mistakes would get copied too. Would it be sufficeint / better to test the format and that the hash changes when there is a change in one of the affected file properties? 

This would mean we get individual tests, one for each property that affects th cache key.

Reply:
The main idea was to not just have hash values hardcoded in the tests. Instead, one can provide the values that are expected and derive the hash from that. Just checking for changed hashes is, in my opinion, less specific than asserting what kind of values should be used in the hash generation.

The risk that this piece of code has a bug in the hash generation that is not spotted is minimal. It is clear that all values are written to the binary stream so the only error left is that the hash is not calculated for the whole memoryStream. But this case is covered by other test that would break if we supplied a constant hash.

Reply:
mhm, okay.
So..compromise: in the test, please add an assert on the expected value (i.e. in this method here) that the Hash string has the expected length and a regex-test for the format. That way, we catch if there's a bug in the test and the hash string ends up being empty or not correctly formatted. 

I know that from reading the code, we can be pretty sure that it's correct. I don't trust mental verification if I can to it in code, though, and I like to follow the rule of not doing the same algorithm in the tests we do in production code. That's how copy&paste errors are introduced. 

Also, test code should be so simple that it's easily verified by reading it, which is why I suggested we drop uneeded details. I can live with being more precise once we add the trip wires. Just for completeness sake on my reasoning: the actual algorithm or order of items added to the hash value isn't important for the unit test. We only want a hashvalue that's collission free. If it's MD5 or SHA256 isn't somehting that's an important detail of the implemetnation since we don't need to do cryptographic safety, only collision-free-ness.